File: leetcode/problems/new_21_game/solution.cpp
class Solution {
public:
    double new21Game(int n, int k, int maxPts) {
        if(k == 0) {
            return 1;
        }
        vector<double> f(k+maxPts, 1.0);
        // f[i] = probability to reach score i
        // f[0] = 1
        // f[1] = f[0]/m;
        // f[2] = f[1]/m + f[0]/m;
        // f[3] = f[2]/m + f[1]/m + f[0]/m;
        // f[i] = (sum for each j in 1..m (f[j]))/m; skip if j>k
        double sum = 1.0;
        for(int i = 1;i<f.size();i++) {
            int j = i-maxPts-1;
            if(j >= 0) {
                sum -= f[j];
            }
            f[i] = sum/maxPts;
            if(i < k) {
                sum += f[i];
            }
        }
        double ret = 1.0;
        for(int i = n+1;i<f.size();i++) {
            ret -= f[i];
        }
        return ret;
    }
};

File: leetcode/problems/new_21_game/solution.rs
impl Solution {
    pub fn new21_game(n: i32, k: i32, max_pts: i32) -> f64 {
        if(k == 0) {
            return 1.0;
        }
        let max_pts = max_pts as usize;
        let k = k as usize;
        let n = n as usize;
        let mut f = vec![1.0;max_pts+k];
        let mut sum = 1.0;
        for i in 1..f.len() {
            if(i > max_pts) {
                sum -= f[i-max_pts-1];
            }
            f[i] = sum/max_pts as f64;
            if(i < k) {
                sum += f[i];
            }
        }
        let x: f64 = f.iter().skip(n + 1).sum();
        return 1.0 - x;
    }
}